### PR TITLE
[Flex][RL] add custom mask support

### PR DESCRIPTION
**Pull request by liangel-02**
adding custom mask support (ie sliding window attention) for Flex Attention RL loop.

- vLLM takes in custom mask mod via `self.vllm_attn.logical_mask_mod` so we just need to pass this through
- in order to work with cuda graphs, we need to cache mask mod generation on generator side in titan so that vLLM recognizes it as the same object and doesn't rebuild
- on vLLM side, we need to move custom mask handling out of the fwd https://github.com/vllm-project/vllm/pull/45232
- unified flex configs to be more clean 

batch invariant grpo script
Screenshot 2026-06-11 at 12 57 56 AM

sliding window unit test 
Screenshot 2026-06-11 at 12 55 10 AM

custom mask mod unit test
Screenshot 2026-06-11 at 12 38 40 PM

### CODE REVIEW

File: torchtitan/models/common/attention.py [hunk 1]
```diff
@@ -180,6 +187,7 @@ class FlexAttention(Module):
     class Config(Module.Config):
         block_size: int | tuple[int, int] = _DEFAULT_SPARSE_BLOCK_SIZE
         kernel_options: dict = field(default_factory=dict)
+        mask_mod: MaskMod.Config | None = None
```
Comment (fegin) [inline torchtitan/models/common/attention.py:190]: We have another `mask_type` in BaseAttention. Having two configs for one thing is not a good idea. This is also implicit. For example, I'm developing a selected gather to improve CP + sliding window. How do I figure out which one should I look into to decide whether I should apply selected gather? `mask_type` or `mask_mod`?

Comment (tianyu-l) [reply torchtitan/models/common/attention.py:190]: fwiw I've killed the `mask_type` in BaseAttention.

File: torchtitan/experiments/rl/models/vllm_wrapper.py [hunk 2]
```diff
@@ -151,21 +152,25 @@ def __init__(
...
+        new_layers = []
+        for layer_cfg in model_config.layers:
+            inner = layer_cfg.attention.inner_attention
+            vllm_backend = VLLMAttentionWrapper.Config(
+                hidden_size=model_config.dim,
+                num_heads=n_heads,
+                num_kv_heads=n_kv_heads,
+                head_dim=head_dim,
+                mask_mod=getattr(inner, "mask_mod", None),
+                sliding_window=sliding_window,
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/models/vllm_wrapper.py:164]: this makes every single layer using the same sliding window, which doesn't make sense in alternating full / sliding window setup? e.g. gpt-oss

File: torchtitan/experiments/rl/models/attention.py [hunk 3]
```diff
@@ -278,9 +283,13 @@ def __init__(self, config: Config) -> None:
             num_kv_heads=num_kv_heads,
             cache_config=cache_config,
             quant_config=None,
+            per_layer_sliding_window=config.sliding_window,
             prefix=f"model.layers.{layer_id}.attention.inner_attention",
         )
 
+        if mask_mod_cfg is not None and config.sliding_window is None:
+            self.vllm_attn.logical_mask_mod = mask_mod_cfg.build().get_mask_mod()
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/models/attention.py:291]: hmm I think we need some discussion here before we build the solution:
- it seems for vllm, we just need to specify the sliding window part, and vllm will handle continuous batching
- but for trainer, we ourselves need to specify "document mask" AND "sliding window mask"
- I don't want to keep two different solutions, `get_attention_masks` for training, and `MaskMod.Config` for inference. I want to unify the solution as much as possible.

File: torchtitan/experiments/rl/models/attention.py [hunk 4]
```diff
@@ -278,9 +283,13 @@ def __init__(self, config: Config) -> None:
             num_kv_heads=num_kv_heads,
             cache_config=cache_config,
             quant_config=None,
+            per_layer_sliding_window=config.sliding_window,
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/models/attention.py:286]: It seems this is taking precedence over mask_mod. Maybe you could take it from https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/gpt_oss/model.py#L48.

https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/gpt_oss/model.py#L128 is layer-specific. To avoid using the same sliding_window size for all layers, we probably should:
- make gpt-oss `Attention.sliding_window_size` optional, when it's None, this `per_layer_sliding_window` should be None.
- remove `use_sliding_attention` and depend on sliding_window_size None or not to determine which mask to use in training.
- still don't accept custom mask mod, because the model we support so far doesn't require beyond sliding window attention.
- (separate) support gpt-oss training with VarlenAttention backend. Basically this is not true https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/gpt_oss/model.py#L66-L68?

File: torchtitan/experiments/rl/config_registry.py [hunk 5]
```diff
@@ -58,18 +67,24 @@ class BatchInvariantFlexConverter(ModelConfigConverter):
...
             inner = layer_cfg.attention.inner_attention
-            if isinstance(inner, FlexAttention.Config):
+            if not isinstance(inner, FlexAttention.Config):
+                continue
+            if self.batch_invariant:
                 inner.kernel_options["BACKEND"] = "TRITON"
                 inner.kernel_options["BLOCK_M"] = self._BLOCK_M
                 inner.kernel_options["BLOCK_N"] = self._BLOCK_N
+            if self.mask_mod is not None:
+                inner.mask_mod = self.mask_mod
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/config_registry.py:87]: this is potentially changing model definition, should not do this.

File: torchtitan/experiments/rl/models/attention.py [hunk 6]
```diff
@@ -226,6 +226,9 @@ class Config(Module.Config):
         num_kv_heads: int
         head_dim: int
         scale: float | None = None
+        mask_mod: MaskMod.Config | None = None
+        sliding_window: int | None = None
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/models/attention.py:230]: Is there a essential difference (e.g. perf), if I use my custom `mask_mod` to support sliding window attention, instead of using the vllm native `sliding_window` int option.

**Review by tianyu-l — CHANGES_REQUESTED** [review]
Let's slow down a bit.

It seems vllm has native sliding window support. Do we need to support custom masking beyond that? gpt-oss is the only model which uses sliding window / full + causal document masking; all other models are using full causal document masking. Maybe we can support sliding window only, and revisit if we need to support new models with more complicated masking.

We need a robust way to error out if user change the model to use some exotic mask, which won't be captured by today's `VLLMAttentionWrapper`.

File: torchtitan/models/common/attention.py (hunk 1 above)
Comment (fegin) [reply torchtitan/models/common/attention.py:190]: oops, I miss that part.

**Comment by liangel-02** [comment]
closing for https://github.com/pytorch/torchtitan/pull/3687
